### ISSUE TITLE
apply function causes performance bottleneck

### ISSUE THREAD

**Issue by 401qingkong** [pytorch/pytorch#90687]
### 🐛 Describe the bug

I implements a pythoch extension cuda operator  for accelerateing softmax computation. And I  test from the kernel level ，Performance improved by 10%. But when I wrapped it into the python front end， there is a performance degradation comparing with the torch softmax implementation.
Now I have located the performance bottleneck to the forward apply function。The code just as follows:
```
class _fuseSoftmax_function(Function):

    @staticmethod
    def forward(ctx, input, half_to_float, dim):
        forward_data= op.softmax_forward(input, dim, half_to_float)
        # ctx.save_for_backward(input, forward_data, torch.Tensor([dim]))
        ctx.save_for_backward(input, forward_data)
        ctx.dim = dim
        return forward_data


    @staticmethod
    def backward(ctx, gradOutput):
        save_input, save_output = ctx.saved_tensors
        grad_input_return = op.softmax_backward(
            gradOutput,
            save_output,
            ctx.dim,
            save_input)
        return grad_input_return, None, None

class FuseSoftmax(torch.nn.Module):
    dim: Optional[int]
    def __init__(self, half_to_float = False, dim = None) -> None:
        super(FuseSoftmax, self).__init__()
        self.half_to_float = half_to_float
        self.dim = dim

    def forward(self, input, **kargs):
        assert input.dim() == 2 or input.dim() == 1, 'Softmax1d requires a 1D or 2D tensor as input'
        if self.dim is None:
            self.dim = _get_softmax_dim("softmax", input.dim(), 5)
        return _fuseSoftmax_function.apply(input, self.half_to_float, self.dim)
```
        

when I change the apply function to `return op.softmax_forward(input, self.dim, self.half_to_float)` , the performance could remains the same as the kernel side and  will get a significant performance improvement.  
What is the apply function doing besides calling the fuseop  forward function. How to reduce redundant time? 
Hope for your reply.

### Versions

centos7.6
pytorch 1.10

cc @ngimel @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

**Comment by albanD** [comment]
Hi,

This is handling all of autograd. So there is quite a bit of logic happening.
One thing you can try is to use to define it on the c++ side to reduce a little bit the overhead, it looks pretty much the same as in python, you can see an example here: https://github.com/pytorch/vision/blob/32d254bbfcf14975f846765775584e61ef25a5bc/torchvision/csrc/ops/autograd/ps_roi_pool_kernel.cpp#L11-L63

**Comment by 401qingkong** [comment]
@albanD 
Thank you for your reply.   I have try to define the new op in the c++ side and I  can compile it successfully.  I test it in a case,  coredump occures when I import the new module.   I notice the op in the example have both the cuda and cpu backend impl. And the op will judge the real input type to choose which op backend  will be called here:
[https://github.com/pytorch/vision/blob/32d254bbfcf14975f846765775584e61ef25a5bc/torchvision/csrc/ops/ps_roi_pool.cpp#L10-L21](url)
My softmax  just have the cuda impl. Does the new bug have anything to do with it？Do I have to define a cpu implementation？Hopes to get more information about how to define a op with its autograd in the c++ side.
